cobalt/watchdog/watchdog.cc

```diff
@@ -518,39 +518,89 @@ bool Watchdog::Ping(const std::string& name, const std::string& info) {
   return client_exists;
 }
 
-std::string Watchdog::GetWatchdogViolations(bool clear) {
+std::string Watchdog::GetWatchdogViolations(const char* clients[],
+                                            int client_num, bool clear) {
   // Gets a json string containing the Watchdog violations since the last
   // call (up to the kWatchdogMaxViolations limit).
-
   if (is_disabled_) return "";
 
   std::string watchdog_json = "";
+  std::string watchdog_json_filtered = "";
 
   starboard::ScopedLock scoped_lock(mutex_);
 
   if (pending_write_) WriteWatchdogViolations();
 
-  starboard::ScopedFile read_file(GetWatchdogFilePath().c_str(),
-                                  kSbFileOpenOnly | kSbFileRead);
-  if (read_file.IsValid()) {
-    int64_t kFileSize = read_file.GetSize();
+  starboard::ScopedFile file(GetWatchdogFilePath().c_str(),
+                             kSbFileOpenOnly | kSbFileRead | kSbFileWrite);
+  if (file.IsValid()) {
+    int64_t kFileSize = file.GetSize();
     std::vector<char> buffer(kFileSize + 1, 0);
-    read_file.ReadAll(buffer.data(), kFileSize);
+    file.ReadAll(buffer.data(), kFileSize);
     watchdog_json = std::string(buffer.data());
 
-    // Removes all Watchdog violations.
-    if (clear) {
-      if (violations_map_) {
-        static_cast<base::DictionaryValue*>(violations_map_.get())->Clear();
-        violations_count_ = 0;
+    // Get all Watchdog violations in clients is given.
+    if (client_num == 0) {
+      // Removes all Watchdog violations.
+      if (clear) {
+        if (violations_map_) {
+          static_cast<base::DictionaryValue*>(violations_map_.get())->Clear();
+          violations_count_ = 0;
+        }
+        starboard::SbFileDeleteRecursive(GetWatchdogFilePath().c_str(), true);
+      }
+      watchdog_json_filtered = watchdog_json;
+    } else {
```

Comment:
So all of the violations should be both in-memory in the violations_map_ and in disk in the json file. The two should always be in sync with the in-memory being more up to date (this is why at the beginning of this function WriteWatchdogViolations() is called to bring the in disk json file to parity with what's in memory).

Previously since it was a simple fetch of an entire file it was just easier to read the entire json file as a string and return that. Here since there's json manipulation it will be much simpler and easier and more efficient to ignore the json file on disk. Just read from the in-memory violations_map_ and manipulate it as needed. Then when you're done trigger a WriteWatchdogViolations().

Reply:
Sure, I can do that if "The two should always be in sync with the in-memory being more up to date" is the case. Just want to double confirm even when json file is the violation from the previous app run, the violation_map_ will be synced before we call getWatchdogViolations()?

Reply:
Well, I think the above statement is not always true. If I only rely on the in_memory map to fetch violations, this unit test will fail: ViolationsJsonShouldPersistAndBeValid. In that test, the violation only exists in json file not in the in memory map. Is this unit test testing an valid case?

Reply:
sgtm, we could move `InitializeViolationsMap` from `UpdateViolationsMap` to the constructor but that will mean we will always force a read... which happens anyways with kabuki calls to `GetWatchdogViolations`

i think it's worth a shot to try moving the code just because messing with the json in memory is much neater than on disk though ultimately the read/write differences will be negligible.